src/main/java/aztech/modern_industrialization/machines/multiblocks/MultiblockMachineBlockEntity.java

```diff
@@ -45,7 +47,61 @@ public boolean isShapeValid() {
         return shapeValid.shapeValid;
     }
 
-    public abstract void unlink();
+    public ShapeMatcher getShapeMatcher() {
+        return shapeMatcher;
+    }
+
+    public ShapeMatcher createShapeMatcher() {
+        return new ShapeMatcher(level, worldPosition, orientation.facingDirection, getActiveShape());
+    }
+
+    protected void onLink() {
+    }
+
+    protected void onUnlink() {
+    }
+
+    protected void onRematch() {
+    }
+
+    protected void onMatchSuccessful() {
+    }
+
+    protected void onMatchFailure() {
+    }
+
+    protected void link() {
+        if (shapeMatcher == null) {
+            shapeMatcher = createShapeMatcher();
+            shapeMatcher.registerListeners(level);
+            onLink();
```

Comment:
Current MI doesn't need this. Do you?

Reply:
Yes, I use it to use a custom shape matcher for my tesla tower to make it only form the multiblock if all energy hatches are the same tier to prevent converting energy tiers when transmitting wirelessly

Reply:
So how are you doing that? Replacing the `shapeMatcher` in `onLink()` makes reasoning about the code too hard, so I hope that's not what you're doing.

Reply:
No, I use the createShapeMatcher method I added :p

Reply:
So you don't use `onLink`? Let's get rid of it then.

Reply:
I do use onLink and onUnlink. I use it to register and unregister additional listeners in the same way MI uses it. Would you prefer I get rid of these two methods and instead made a custom ShapeMatcher that adds to the un/registerListeners methods?

Reply:
Yes let's do that